src/sbvr-api/actions.ts

```diff
@@ -0,0 +1,141 @@
+import type { ODataQuery } from '@balena/odata-parser';
+import {
+	BadRequestError,
+	type ParsedODataRequest,
+} from '../sbvr-api/uri-parser.js';
+import { api, type Response } from '../sbvr-api/sbvr-utils.js';
+import type { Tx } from '../database-layer/db.js';
+import { sbvrUtils } from '../server-glue/module.js';
+import type { AnyObject, Params } from 'pinejs-client-core';
+import { UnauthorizedError } from '../sbvr-api/errors.js';
+
+export type ODataActionRequest = Omit<ParsedODataRequest, 'odataQuery'> & {
+	odataQuery: Omit<ODataQuery, 'property'> & {
+		property: {
+			resource: string;
+		};
+	};
+};
+
+type ActionReq = Express.Request;
+
+export type ODataActionArgs<Vocab extends string> = {
+	request: ODataActionRequest;
+	tx: Tx;
+	api: (typeof api)[Vocab];
+	id: Params['id'];
+	req: ActionReq;
+};
+export type ODataAction<Vocab extends string = string> = (
+	args: ODataActionArgs<Vocab>,
+) => Promise<Response>;
+
+const actions = {} as {
+	[vocab: string]: {
+		[resourceName: string]: {
+			[actionName: string]: ODataAction;
+		};
+	};
+};
+
+export const isActionRequest = (
+	request: ParsedODataRequest,
+): request is ODataActionRequest => {
+	// OData actions must always be POST
+	// See: https://www.odata.org/blog/actions-in-odata/
+	return (
+		request.method === 'POST' &&
+		request.odataQuery.property?.resource != null &&
+		actions?.[request.vocabulary]?.[request.resourceName]?.[
+			request.odataQuery.property.resource
+		] != null
+	);
+};
+
+export const runAction = async (
+	request: ODataActionRequest,
+	req: ActionReq,
+): Promise<Response> => {
+	if (api[request.vocabulary] == null) {
+		throw new BadRequestError();
+	}
+
+	const actionName = request.odataQuery.property.resource;
+	const action =
+		actions?.[request.vocabulary]?.[request.resourceName]?.[actionName];
+	if (!action) {
+		throw new BadRequestError();
+	}
+
+	return sbvrUtils.db.transaction(async (tx) => {
+		// in practice, the parser does not currently allow actions without a key
+		// e.g. /vocab/resource(...)/action compiles but
+		// /vocab/resource/action won't even compile, so (current) reality is that if canAcccess
+		// does not throw, there will be an id, but I kept the interface as optional id
+		// for when the parser has support for root level actions
+		const id = request.odataQuery.key
+			? await canRunAction(request, req, actionName, tx)
+			: undefined;
+
+		const applicationApi = api[request.vocabulary].clone({
+			passthrough: { tx, req },
+		});
+
+		return await action({
+			request: request,
+			tx,
+			api: applicationApi,
+			req: req,
+			id,
+		});
+	});
+};
+
+export const addAction = <Vocab extends string>(
```

Comment:
Different than addHook we do not pass the method (as per odata definition actions are always POST) - however, wen need to pass a name